pkg/kademlia/routing_helpers_test.go

```diff
@@ -63,193 +64,227 @@ func mockNode(id string) *pb.Node {
+			added:  false,
+			kadIDs: [][]byte{[]byte{63, 255}, []byte{79, 255}, []byte{95, 255}, []byte{127, 255}, []byte{255, 255}},
+			nodeIDs: [][]string{[]string{"9O", ":O", ";O", "=O", ">O", "?O"},
+				[]string{"LO", "MO", "NO", "OO"},
+				[]string{"PO", "QO", "SO"},
+				[]string{},
+				[]string{},
+			},
+		},
+		{testID: "KO",
+			node:   mockNode("KO"),
+			added:  true,
+			kadIDs: [][]byte{[]byte{63, 255}, []byte{79, 255}, []byte{95, 255}, []byte{127, 255}, []byte{255, 255}},
+			nodeIDs: [][]string{[]string{"9O", ":O", ";O", "=O", ">O", "?O"},
+				[]string{"KO", "LO", "MO", "NO", "OO"},
+				[]string{"PO", "QO", "SO"},
+				[]string{},
+				[]string{},
+			},
+		},
+		{testID: "JO",
+			node:   mockNode("JO"),
+			added:  true,
+			kadIDs: [][]byte{[]byte{63, 255}, []byte{79, 255}, []byte{95, 255}, []byte{127, 255}, []byte{255, 255}},
+			nodeIDs: [][]string{[]string{"9O", ":O", ";O", "=O", ">O", "?O"},
+				[]string{"JO", "KO", "LO", "MO", "NO", "OO"},
+				[]string{"PO", "QO", "SO"},
+				[]string{},
+				[]string{},
+			},
+		},
+		{testID: "]O",
+			node:   mockNode("]O"),
+			added:  true,
+			kadIDs: [][]byte{[]byte{63, 255}, []byte{79, 255}, []byte{95, 255}, []byte{127, 255}, []byte{255, 255}},
+			nodeIDs: [][]string{[]string{"9O", ":O", ";O", "=O", ">O", "?O"},
+				[]string{"JO", "KO", "LO", "MO", "NO", "OO"},
+				[]string{"PO", "QO", "SO", "]O"},
+				[]string{},
+				[]string{},
+			},
+		},
+		{testID: "^O",
+			node:   mockNode("^O"),
+			added:  true,
+			kadIDs: [][]byte{[]byte{63, 255}, []byte{79, 255}, []byte{95, 255}, []byte{127, 255}, []byte{255, 255}},
+			nodeIDs: [][]string{[]string{"9O", ":O", ";O", "=O", ">O", "?O"},
+				[]string{"JO", "KO", "LO", "MO", "NO", "OO"},
+				[]string{"PO", "QO", "SO", "]O", "^O"},
+				[]string{},
+				[]string{},
+			},
+		},
+		{testID: "_O",
+			node:   mockNode("_O"),
+			added:  true,
+			kadIDs: [][]byte{[]byte{63, 255}, []byte{79, 255}, []byte{95, 255}, []byte{127, 255}, []byte{255, 255}},
+			nodeIDs: [][]string{[]string{"9O", ":O", ";O", "=O", ">O", "?O"},
+				[]string{"JO", "KO", "LO", "MO", "NO", "OO"},
+				[]string{"PO", "QO", "SO", "]O", "^O", "_O"},
+				[]string{},
+				[]string{},
+			},
+		},
+		{testID: "@O: split bucket 2",
+			node:   mockNode("@O"),
+			added:  true,
+			kadIDs: [][]byte{[]byte{63, 255}, []byte{71, 255}, []byte{79, 255}, []byte{95, 255}, []byte{127, 255}, []byte{255, 255}},
+			nodeIDs: [][]string{[]string{"9O", ":O", ";O", "=O", ">O", "?O"},
+				[]string{"@O"},
+				[]string{"JO", "KO", "LO", "MO", "NO", "OO"},
+				[]string{"PO", "QO", "SO", "]O", "^O", "_O"},
+				[]string{},
+				[]string{},
+			},
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.testID, func(t *testing.T) {
+			ok, err := rt.addNode(c.node)
+			assert.Equal(t, c.added, ok)
+			assert.NoError(t, err)
+			kadKeys, err := rt.kadBucketDB.List(nil, 0)
+			assert.NoError(t, err)
+			for i, v := range kadKeys {
+				assert.Equal(t, storage.Key(c.kadIDs[i]), v)
```

Comment:
are kadBucketDB.List and getNodeIDsWithinKBucket required to be in sorted order?

Reply:
they will be based on how bolt is organized

Reply:
I'm asking whether it's required to be that way. Let's say we create `kadBucketDB` to use a `map[]` instead which has non-deterministic order, would routing still work with that?

If it would work, then this test is overly restrictive in what it expects from `kadBucketDB`, meaning a potential failure when substituting the implementation. In essence the "order of keys" might be an implementation detail.